numpy/core/src/multiarray/methods.c

```diff
@@ -2694,6 +2694,166 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
     return c;
 }
 
+
+static void
+array_dlpack_deleter(DLManagedTensor *self)
+{
+    PyArrayObject *array = (PyArrayObject *)self->manager_ctx;
+    // This will also free the strides as it's one allocation.
+    PyMem_Free(self->dl_tensor.shape);
+    PyMem_Free(self);
+    Py_XDECREF(array);
+}
+
+/* This is exactly as mandated by dlpack */
+static void array_dlpack_capsule_deleter(PyObject *self)
+{
+    if (PyCapsule_IsValid(self, NPY_DLPACK_USED_CAPSULE_NAME)) {
+        return;
+    }
+    DLManagedTensor *managed = 
+        (DLManagedTensor *)PyCapsule_GetPointer(self, NPY_DLPACK_CAPSULE_NAME);
+    if (managed == NULL) {
+        return;
+    }
+    managed->deleter(managed);
+}
```

Comment:
Getting this function right is hard - I've updated my implementation in https://github.com/dmlc/dlpack/issues/51#issuecomment-849794854, but it may need further iteration.

The fact that the glue between dlpack and CPython is hard to get right suggests to me that this should really be handled by a dependency written by the dlpack maintainers, rather than burdening every downstream implementer which writing their own version.

Reply:
We just have to get things right for the cases that we produce. If/when DLPack adds a standard implementation, I'll replace the one here.

Reply:
I opened an issue (https://github.com/dmlc/dlpack/issues/74) to discuss about `deleter` and other possible DLPack C APIs.

Reply:
@hameerabbasi, you still need the PyErr_Fetch stuff from the comment I link